### ISSUE TITLE
Soldier runtime takes a very long time to start

### ISSUE THREAD

**Issue by rbernon** [ValveSoftware/steam-runtime#323]
Using Soldier 0.20201124.0 on Debian Unstable, with AMD e-GPU and custom Mesa build installed in $HOME.

```shell
bwrap --version
bubblewrap 0.4.1
```

It makes every game startup horribly slow and it's reproducible by running simple commands directly:

```shell
time ~/Games/SteamLinuxRuntime_soldier/run ls
real	0m55.230s
user	0m40.199s
sys	0m15.071s
```

Timestamped log file of `env -i ~/Games/SteamLinuxRuntime_soldier/run ls`: [steam-runtime-soldier-ls.log](https://github.com/ValveSoftware/steam-runtime/files/5683124/steam-runtime-soldier-ls.log)

The culprit seems to be all the `bwrap` calls that already take time individually, but also that add up with all the vulkan devices and layers checks.

**Comment by smcv** [comment]
The setup does take a while, but I haven't seen it be this slow on our test machines! The timing I've seen is more like 10 seconds than 55, and my usual test hardware is a 9-year-old laptop with a magnetic HDD rather than SSD, which I'd assumed would be a worst case scenario for performance. Is there anything unusual about your test machine that would make it slower?

At the moment our focus is on making it reliable rather than on making it fast, so this issue is likely to stay open for a while. If we make it reliable first, then we'll know what we need it to do, and we can make it faster afterwards. If we try to make it fast first, we'll probably have to undo many of the optimizations that made it faster in order to make it reliable, so the end result will just be that we wasted time on optimizing the wrong thing.

> The culprit seems to be all the `bwrap` calls that already take time individually, but also that add up with all the vulkan devices and layers checks.

These are necessary: they're how we make the container functional. We can't use the graphics drivers from the host system without doing this work. When we have the container otherwise reliable, we might be able to speed them up by coalescing them into fewer, larger batches, or by having a cache, but we can't eliminate this work completely.

The time taken might be in `bwrap` itself, or it might be in `capsule-capture-libs`. Please could you get a similar log with `PRESSURE_VESSEL_COPY_RUNTIME_INTO="$HOME/Games/SteamLinuxRuntime_soldier/var"` on the same slow system? If that's significantly faster, then that probably points to `bwrap` as the "expensive" part (which we can potentially avoid even in the relatively short term by adjusting how we do things). If not, then it's more likely `capsule-capture-libs` being the problem (which, realistically, we can't avoid, so we would have to figure out ways to speed it up instead).

It looks as though you're seeing times of 300-500ms per `bwrap`/`capsule-capture-libs` pair, whether it captures one library or many, suggesting that the fixed cost of `bwrap` and starting `capsule-capture-libs` is more influential than the incremental cost of each library we deal with. If that's the case, then there might be some mileage in coalescing calls into batches (so we try to do all the DRI drivers in one go, for example).

I'm very reluctant to use a cache at this stage in development, because where you have caches, you also have a need for cache invalidation, widely reputed to be one of the two hardest things in computer science (along with naming and off-by-one errors). If it turns out we need a cache, we should only be trying that when the rest of the reliability issues are solved.

Similarly, if we're CPU-bound, parallelizing might help; but while we're still dealing with other reliability issues, not having to think about concurrency is really valuable.

**Comment by rbernon** [comment]
> The setup does take a while, but I haven't seen it be this slow on our test machines! The timing I've seen is more like 10 seconds than 55, and my usual test hardware is a 9-year-old laptop with a magnetic HDD rather than SSD, which I'd assumed would be a worst case scenario for performance. Is there anything unusual about your test machine that would make it slower?

Nothing that I know of, it's not that old and it has a SSD too. It's not happening on my work laptop, which has mostly the same setup, either with `soldier 0.20201203.0` or `soldier 0.20201104.0`. I'll try to get more logs.

The main trouble was also that it is doing the "Preparing..." setup every time the game was started (Kingdom Come Deliverance in this case). As you say there is no cache it's may be supposed to be like that, but it may also be an issue on its own.

**Comment by smcv** [comment]
> The main trouble was also that it is doing the "Preparing..." setup every time the game was started (Kingdom Come Deliverance in this case). As you say there is no cache it's may be supposed to be like that

For the moment, yes, we do all the setup every time. There are two layers involved.

One layer of setup is Steam's "install scripts", which install the game's dependencies if necessary (DirectX, C++ standard library, that sort of thing). I think this is what makes the Steam client show `Preparing...`. Doing that every time and not just on first-run is a Steam client decision, and pressure-vessel is not in a position to change that.

The other layer of setup is that pressure-vessel needs to set up its container before we can do anything inside the container. If there are "install scripts", we do this the first time Steam starts one of those, which means the delay inside pressure-vessel happens while Steam is showing `Preparing...`, before the first time we hand over control to the `proton` script. If there were no install scripts, in principle we'd do the container setup as a side-effect of starting the actual game, just before handing over control to `proton waitforexitandrun` (but I think in practice Steam might always be running an install script, even if it's a trivial one). At the moment we do this every time instead of caching the results of inspecting the host OS, because if we had a cache, we would need to be able to invalidate it every time there is a change to the host OS' graphics driver and library stacks or to the environment variables that affect which libraries we choose, and it's far from obvious how to do that in a portable way.

**Comment by rbernon** [comment]
So I've run the command with `PRESSURE_VESSEL_COPY_RUNTIME_INTO` and the issue is still present.

As you said it was possibly then `capsule-capture-libs` fault, I've then replaced `i386-linux-gnu-capsule-capture-libs` and `x86_64-linux-gnu-capsule-capture-libs` with empty scripts, and it still takes 500ms for every call.

As far as I can tell g_spawn_sync is taking a long time to return (or to start), regardless of what is then executed. With strace I can see that `pressure-vessel-wrap`, calls `fcntl(fd, F_SETFD, FD_CLOEXEC)` for all possible fd values, up to `RLIMIT_NOFILE`.

In my case, I have `DefaultLimitNOFILE=1048576`, probably since some Wine esync setup, and this is why it is taking so long. Reducing the limit to 4096 for instance, fixes the issue and the execution completes in a reasonable time.

**Comment by rbernon** [comment]
I believe many people that used to play games with Wine and esync may have a similar setup, because it was advised at some point to increase this limit a lot for esync compatibility. I don't know if it's required anymore, but that may have stayed since then like it was for me.

**Comment by smcv** [comment]
> With strace I can see that pressure-vessel-wrap, calls fcntl(fd, F_SETFD, FD_CLOEXEC) for all possible fd values, up to RLIMIT_NOFILE.

Aha! That is something we can probably solve without making pressure-vessel harder to develop, for example by backporting modern GLib changes into scout's GLib (which pressure-vessel bundles a copy of, to be able to run anywhere, without having non-glibc system library dependencies).

If you definitely have both 32- and 64-bit system copies of GLib/GObject/GIO, deleting the bundled copies would probably work as a workaround. (At your own risk, etc.)

**Comment by smcv** [comment]
Workaround: set the soft rlimit to be small (1024 is recommended), and the hard limit to be large (a million or so), as is now recommended in <https://github.com/zfigura/wine/blob/esync/README.esync>. This still lets you use a million fds for esync, but limits processes that don't opt-in to the extra-high limit to the smaller soft limit. It also means `select()` will definitely work.

**Comment by smcv** [comment]
> If you definitely have both 32- and 64-bit system copies of GLib/GObject/GIO, deleting the bundled copies would probably work as a workaround.

It doesn't. We bypass GLib for this and use our own code (actually taken from Flatpak), to work around a different problem with older GLib.

https://gitlab.steamos.cloud/steamrt/steam-runtime-tools/-/merge_requests/204 solves the vast majority of this. There's probably more that could be done, but I'd prefer to defer that until we have pressure-vessel working on more systems.

**Comment by rbernon** [comment]
I'm closing this as there's a workaround, and IIUC some kind of a fix too.

**Comment by smcv** [comment]
This is mostly fixed in pressure-vessel versions >= 0.20210105.0: having a soft fd limit of a million fds now increases startup time by something like 20%, rather than 900%.

These versions are available in the beta branch of SteamLinuxRuntime and SteamLinuxRuntime_soldier starting from today (check VERSIONS.txt). They are not in the normal/stable branch yet, but probably will be soon.

I would still recommend keeping the system-wide soft limit for RLIMIT_NOFILE to the default 1024, with the hard limit set to 512K (the default in systemd >= 240) or 1024K (as suggested by README.esync), because this is not the only place where a large soft limit has bad effects. See the systemd 240 NEWS for more details. There is nothing systemd-specific about this, and setting those limits is equally good for users of other init systems.
